:seedling: Add SetControlPlaneEndpoint function

razashahid107 opened the pull request:
<!-- If this is your first PR, welcome! Please make sure you read the [contributing guidelines](../CONTRIBUTING.md#contributing-a-patch). -->
<!-- please add an icon to the title of this PR and delete this line and similar ones -->
<!-- the icon will be either ⚠️ (:warning:, major or breaking changes), ✨ (:sparkles:, feature additions), 🐛 (:bug:, patches and bugfixes), 📖 (:book:, documentation or proposals), or 🌱 (:seedling:, minor or other) -->

**What this PR does / why we need it**:
The following PR addresses the issue #977, and aims to add the function SetControlPlaneEndpoint() and unit tests for it. 
**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #977 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._
No it does not

**TODOs**:

- [ ] squash commits
- [ ] include documentation
- [ ] add unit tests

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -848,90 +848,102 @@ func TestSetControlPlaneEndpoint(t *testing.T) {
 		}
 		if (hetznerCluster.Spec.ControlPlaneEndpoint != nil){
-			t.Fatal("Wrong value for hetznerCluster.Spec.ControlPlaneEndpoint")
+			t.Fatalf("Wrong value for hetznerCluster.Spec.ControlPlaneEndpoint. Got: %s, Want: '<nil>'", hetznerCluster.Spec.ControlPlaneEndpoint)
```
Want: nil. In this case we actually want nil, not the string "<nil>" that we have as well in the code

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -848,90 +848,102 @@ func TestSetControlPlaneEndpoint(t *testing.T) {
 		}
 		if (hetznerCluster.Spec.ControlPlaneEndpoint == nil){
-			t.Fatal("ControlPlaneEndpoint should not be nil")
+			t.Fatalf("ControlPlaneEndpoint should not be nil. Got: %s, Want: (Host: 'xyz')", hetznerCluster.Spec.ControlPlaneEndpoint)
```
here you can make it easier: "ControlPlaneEndpoint must not be nil" without anything else, as you had before

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -848,90 +848,102 @@ func TestSetControlPlaneEndpoint(t *testing.T) {
 		}
 		if (hetznerCluster.Spec.ControlPlaneEndpoint == nil){
-			t.Fatal("hetznerCluster.Spec.ControlPlaneEndpoint should not be nil")
+			t.Fatalf("Wrong input for hetznerCluster.Spec.ControlPlaneEndpoint. Got: %s, Want: (Host: 'xyz', Port: 1)", hetznerCluster.Spec.ControlPlaneEndpoint)
```
you can leave it at the original "hetznerCluster.Spec.ControlPlaneEndpoint should not be nil"

razashahid107 commented:
@janiskemper. I've made the required changes. Can you please review again? I've squashed my commits

janiskemper commented on controllers/hetznercluster_controller.go:
```diff
@@ -260,6 +263,34 @@ func (r *HetznerClusterReconciler) reconcileNormal(ctx context.Context, clusterS
 	return reconcile.Result{}, nil
 }
 
+func SetControlPlaneEndpoint(hetznerCluster *infrav1.HetznerCluster) (bool) {
```
you don't need to export the function -> setControlPlaneEndpoint

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
 })
+
+func TestSetControlPlaneEndpoint(t *testing.T) {
+	t.Run("function should return false if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled is not enabled and hetznerCluster.Spec.ControlPlaneEndpoint is nil", func(t *testing.T) {
```
return false and don't make changes to ControlPlaneEndpoint if load balancer is not enabled and ControlPlaneEndpoint nil

razashahid107 replied:
I should just change the name of the test right?

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
+		}
+		assert.Equal(t, false, got)
+	})
+	t.Run("function should return true if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled is false but hetznerCluster.Spec.ControlPlaneEndpoint is not nil", func(t *testing.T) {
```
return false and don't make changes to ControlPlaneEndpoint if load balancer is not enabled and ControlPlaneEndpoint is not nil

razashahid107 replied:
I think this should be true right? Because the last condition would get triggered, so it would return true

janiskemper replied:
Yes, I did a mistake there. Replace false with true

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
+
+func TestSetControlPlaneEndpoint(t *testing.T) {
+	t.Run("function should return false if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled is not enabled and hetznerCluster.Spec.ControlPlaneEndpoint is nil", func(t *testing.T) {
+		hetznerCluster := &infrav1.HetznerCluster{Spec: infrav1.HetznerClusterSpec{ControlPlaneLoadBalancer: infrav1.LoadBalancerSpec{Enabled: false}, ControlPlaneEndpoint: nil}}
```
don't define the struct in one line. take multiple lines

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
+	t.Run("function should return false if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled is not enabled and hetznerCluster.Spec.ControlPlaneEndpoint is nil", func(t *testing.T) {
+		hetznerCluster := &infrav1.HetznerCluster{Spec: infrav1.HetznerClusterSpec{ControlPlaneLoadBalancer: infrav1.LoadBalancerSpec{Enabled: false}, ControlPlaneEndpoint: nil}}
+		got := SetControlPlaneEndpoint(hetznerCluster)
+		if (hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled == true){
```
we never change anything in ControlPlaneLoadBalancer. You really don't have to test it. Focus on changes in ControlPlaneEndpoint. You can remove this

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
+		if (hetznerCluster.Spec.ControlPlaneEndpoint != nil){
+			t.Fatalf("Wrong value for hetznerCluster.Spec.ControlPlaneEndpoint. Got: %s, Want: <nil>", hetznerCluster.Spec.ControlPlaneEndpoint)
+		}
+		assert.Equal(t, false, got)
```
stick to t.Fatalf. Otherwise we import the assert package which we don't really need.

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
 	})
 })
+
+func TestSetControlPlaneEndpoint(t *testing.T) {
```
leave some space between the different lines.. You have not a single empty line here. Leave one between every test, after you define the hetznerCluster, after you start the if statements with the tests, etc.

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
+		if (hetznerCluster.Spec.ControlPlaneEndpoint == nil){
+			t.Fatalf("ControlPlaneEndpoint must not be nil")
+		}
+		if (hetznerCluster.Spec.ControlPlaneEndpoint.Host != "xyz"){
```
you also should test here that the port is still not set... otherwise we cannot be sure that no changes have been made to the port

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
+	t.Run("function should return false if ControlPlaneLoadBalancer is enabled and hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4 is '<nil>'", func(t *testing.T) {
+		hetznerCluster := &infrav1.HetznerCluster{Spec: infrav1.HetznerClusterSpec{ControlPlaneLoadBalancer: infrav1.LoadBalancerSpec{Enabled: true}}, Status: infrav1.HetznerClusterStatus{ControlPlaneLoadBalancer: &infrav1.LoadBalancerStatus{IPv4: "<nil>"}}}
+		got := SetControlPlaneEndpoint(hetznerCluster)
+		if (hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled != true){
```
not important

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
+		if (hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled != true){
+			t.Fatalf("Wrong value for ControlPlaneLoadBalancer.Enabled. Got: %v, Want: true", hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled)
+		}
+		if (hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4 != "<nil>"){
```
also not important. Focus on ControlPlaneEndpoint

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
+		if (hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4 != "<nil>"){
+			t.Fatalf("Wrong value for Status.ControlPlaneLoadBalancer.IPv4. Got: %s, Want: '<nil>'", hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4)
+		}
+		assert.Equal(t, false, got)
```
you tested now that there are no changes in ControlPlaneLoadBalancer which is kinda obvious in setControlPlaneEndpoint function, but you didn't check that no changes have been made to the ControlPlaneEndpoint! 

In this test case you would check that it is still nil. So you should have another test case where you have host or port set and check that both are still the same as well.

razashahid107 replied:
About the second test that you have recommended here, this one is also done below where I check the changes made to ControlPlaneEndpoint (when IPv4 and Port are not nil)

janiskemper replied:
ok. then you can maybe remove it

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
+		}
+		assert.Equal(t, false, got)
+	})
+	t.Run("function should return true if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled is true and hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4 contains a value (not nil)", func(t *testing.T) {
```
why don't you test the changes to ControlPlaneEndpoint in this test as well?

razashahid107 replied:
I think I should just remove this test. The test below checks values of controlplaneEndpoint and is pretty much similar

janiskemper replied:
then do that

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
+		}
+		assert.Equal(t, true, got)
+	})
+	t.Run("function should return true if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled is true, hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4 is not nil and hetznerCluster.Spec.ControlPlaneEndpoint is nil", func(t *testing.T) {
```
in your test description you don't mention the changes to ControlPlaneEndpoint...

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
+		}
+		assert.Equal(t, true, got)
+	})
+	t.Run("function should return true if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled is true, hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4 is not nil, hetznerCluster.Spec.ControlPlaneEndpoint.Host is an empty string and hetznerCluster.Spec.ControlPlaneEndpoint.Port is 0", func(t *testing.T) {
```
same thing. Include the changes (or no changes) that you expect for ControlPlaneEndpoint

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +842,100 @@ var _ = Describe("reconcileRateLimit", func() {
+		}
+		assert.Equal(t, true, got)
+	})
+	t.Run("function should return true if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled is false but hetznerCluster.Spec.ControlPlaneEndpoint is not nil", func(t *testing.T) {
```
you have this test already above I believe

razashahid107 commented:
@janiskemper . I've implemented the recommended changes. Can you please take a look?

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -844,55 +843,71 @@ var _ = Describe("reconcileRateLimit", func() {
+		hetznerCluster := &infrav1.HetznerCluster{
+			Spec: infrav1.HetznerClusterSpec{
+				ControlPlaneLoadBalancer: infrav1.LoadBalancerSpec{Enabled: false}, 
+				ControlPlaneEndpoint: nil},
```
ControlPlaneEndpoint: nil,
},

diffferent line break...

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -844,55 +843,71 @@ var _ = Describe("reconcileRateLimit", func() {
+		hetznerCluster := &infrav1.HetznerCluster{
+			Spec: infrav1.HetznerClusterSpec{
+				ControlPlaneLoadBalancer: infrav1.LoadBalancerSpec{Enabled: false}, 
+				ControlPlaneEndpoint: &clusterv1.APIEndpoint{Host: "xyz"}},
```
you don't have a check for the port

razashahid107 replied:
I had it this way intentionally becuase the requirement is only to not have ControlPlaneEndpoint not nil. So I thought keeping it this way will allow it to pass through.

janiskemper replied:
In this test case you wanted to check that ControlPlaneEndpoint does not change. You had left the port empty before. I believe that was better as now, where you added the port to ControlPlaneEndpoint as well. 

Leave it empty and check in the test that it is left empty!

razashahid107 replied:
Okay I'm a little confused with this one @janiskemper. How do I leave an integer input field empty? Do you mean I should set its value to 0, or i just leave it unmnetioned? If i set it to zero its value will not change since we do not enter that condition in this test. I'm sorry I'm just a little confused

janiskemper replied:
0

janiskemper replied:
and yes, you should test that the 0 is not replaced by anything

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -844,55 +843,71 @@ var _ = Describe("reconcileRateLimit", func() {
-	t.Run("function should return true if hetznerCluster.Spec.ControlPlaneLoadBalancer.Enabled is true and hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4 contains a value (not nil)", func(t *testing.T) {
-		hetznerCluster := &infrav1.HetznerCluster{Spec: infrav1.HetznerClusterSpec{ControlPlaneLoadBalancer: infrav1.LoadBalancerSpec{Enabled: true, Port: 12}}, Status: infrav1.HetznerClusterStatus{ControlPlaneLoadBalancer: &infrav1.LoadBalancerStatus{IPv4: "xyz"}}}
+
+	t.Run("return false if load balancer is enabled but IPv4 is '<nil>'", func(t *testing.T) {
```
and ControlPlaneEndpoint will not be changed

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -844,55 +843,71 @@ var _ = Describe("reconcileRateLimit", func() {
+			Spec: infrav1.HetznerClusterSpec{
+				ControlPlaneLoadBalancer: infrav1.LoadBalancerSpec{
+					Enabled: true, 
+					Port: 11}, 
```
always fix the line break. you never do it this way with structs, you always close it in the next line, i.e.

Port: 11,
},

janiskemper commented on controllers/hetznercluster_controller.go:
```diff
@@ -260,6 +263,34 @@ func (r *HetznerClusterReconciler) reconcileNormal(ctx context.Context, clusterS
 	return reconcile.Result{}, nil
 }
 
+func setControlPlaneEndpoint(hetznerCluster *infrav1.HetznerCluster) bool {
```
add a comment to this function
// setControlPlaneEndpoint updates hetznerCluster.Spec.ControlPlaneEndpoint and returns whether the hetznerCluster is ready or not

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +841,142 @@ var _ = Describe("reconcileRateLimit", func() {
+				ControlPlaneEndpoint: nil,
+			},
+		}
+		got := setControlPlaneEndpoint(hetznerCluster)
```
here for example you could leave an empty line... Please always put an empty line if the blocks of code do not directly relate to each other.

For example: Empty line before and after this line `got := setControlPlaneEndpoint(hetznerCluster)`

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +841,142 @@ var _ = Describe("reconcileRateLimit", func() {
+			},
+		}
+		got := setControlPlaneEndpoint(hetznerCluster)
+		if (hetznerCluster.Spec.ControlPlaneEndpoint != nil){
```
remove ()

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +841,142 @@ var _ = Describe("reconcileRateLimit", func() {
+		}
+	})
+
+	t.Run("return true and don't make changes to ControlPlaneEndpoint if load balancer is not enabled but ControlPlaneEndpoint is not nil", func(t *testing.T) {
```
but -> and

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +841,142 @@ var _ = Describe("reconcileRateLimit", func() {
+		}
+	})
+
+	t.Run("return false if load balancer is enabled but IPv4 is '<nil>'. ControlPlaneEndpoint should not change", func(t *testing.T) {
```
but -> and

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +841,142 @@ var _ = Describe("reconcileRateLimit", func() {
+			},
+		}
+		got := setControlPlaneEndpoint(hetznerCluster)
+		// ControlPlaneEnpdoint should still remain nil
```
remove this. It is obvious from the good error message below

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +841,142 @@ var _ = Describe("reconcileRateLimit", func() {
+		}
+	})
+
+	t.Run("return true if load balancer is enabled and IPv4 is not nil, but ControlPlaneEndpoint is nil. Values of ControlPlaneEndpoint.Host and ControlPlaneEndpoint.Port will get updated", func(t *testing.T) {
```
better message: 
... if load balancer is enabled, IPv4 is not nil, and ControlPlaneEndpoint is nil

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +841,142 @@ var _ = Describe("reconcileRateLimit", func() {
+			},
+		}
+		got := setControlPlaneEndpoint(hetznerCluster)
+		if (hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4 != "xyz"){
```
remove ()

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +841,142 @@ var _ = Describe("reconcileRateLimit", func() {
+		if (hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4 != "xyz"){
+			t.Fatalf("Wrong input for hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4. Got: %s, Want: 'xyz'", hetznerCluster.Status.ControlPlaneLoadBalancer.IPv4)
+		}
+		// value for ControlPlaneEndpoint (Host and Port) will change after SetControlPlaneEndpoint() is run
```
you can remove this comment here in this line

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -840,3 +841,142 @@ var _ = Describe("reconcileRateLimit", func() {
+			},
+		}
+		got := setControlPlaneEndpoint(hetznerCluster)
+		// Values of hetznerCluster.Spec.ControlPlaneEndpoint.Host and hetznerCluster.Spec.ControlPlaneEndpoint.Port should change after execution of the function SetControlPlaneEndpoint()
```
you can remove this comment. You have it above already extensively

janiskemper commented on controllers/hetznercluster_controller.go:
```diff
@@ -260,6 +263,34 @@ func (r *HetznerClusterReconciler) reconcileNormal(ctx context.Context, clusterS
+					Host: defaultHost,
+					Port: defaultPort,
+				}
+				return true
```
I just realized we return true here and in the else statement, so in every case.. We can remove these two return true and have one common one after the if else statement, i.e.
```
if hetznerCluster.Spec.ControlPlaneEndpoint == nil{
} else {
}
return true
```

janiskemper reviewed (changes requested):
to complete the tests, you should have also two more tests that are rather similar to the last one you have, where you already have a host or a port set so that it does not get updates. 

Also you should have a test where both is set already and also doesn't get replaced (maybe by a different value from ControlPlaneLoadBalancer. 

So: you can implement three more tests!

razashahid107 commented:
@janiskemper I've made the changes. I request you take another look

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -902,17 +910,19 @@ func TestSetControlPlaneEndpoint(t *testing.T) {
-		// ControlPlaneEnpdoint should still remain nil
-		if (hetznerCluster.Spec.ControlPlaneEndpoint != nil){
+
+		if hetznerCluster.Spec.ControlPlaneEndpoint != nil{
```
this looks weird.. there should always be a space between nil and {... Is it possible that you don't have linting and formatting in place that fixes stuff like that? Please have a look into that. You can e.g. configure vscode in a way that it formats your code on save.

razashahid107 replied:
You are right. I do use it. However, I thought the code was already formatted for, hence did not do it. I'll make sure I use it before committing next time

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -966,15 +981,116 @@ func TestSetControlPlaneEndpoint(t *testing.T) {
+				},
+			},
+		}
+		got := setControlPlaneEndpoint(hetznerCluster)
```
space before

janiskemper commented on controllers/hetznercluster_controller_test.go:
```diff
@@ -966,15 +981,116 @@ func TestSetControlPlaneEndpoint(t *testing.T) {
+				},
+			},
+		}
+		got := setControlPlaneEndpoint(hetznerCluster)
```
space before

janiskemper commented:
I believe that you haven't implemented the additional tests right?

razashahid107 commented:
@janiskemper I did implement the three additional tests you recommended

janiskemper commented:
true. My bad. You can resolve the two unresolved conversations and then you are done I guess!

razashahid107 commented:
@janiskemper I think we should be good now. I had added a port value for ControlPlaneLoadBalancer.Port too. I think that was unnecessary. I removed that, hence the two commits. 
I've ammended that. Please take a look.

janiskemper commented:
you can squash now

razashahid107 commented:
@janiskemper i've squashed the commits. Learnt so much with this issue. Thank you

janiskemper commented:
did you call `make verify` and `make test`? If so, we can mark it as "ready for review"

janiskemper commented:
thanks @razashahid107 for this PR :)

razashahid107 commented:
I loved doing this one. Got to learn so much. Thank you for being supportive @janiskemper @batistein
